File: phonebook/ContactManager.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Contact {
	String firstName;
	String lastName;
	String phoneNumber;
    
	public Contact(String firstName, String lastName, String phoneNumber) {
		this.firstName = firstName;
		this.lastName = lastName;
		this.phoneNumber = phoneNumber;
	}
    
	public String getFullName() {
		return firstName + " " + lastName;
	}
}

public class ContactManager {
	List<Contact> contacts = new ArrayList<>();
	public void addContact(Contact newContact) {
		if (findByPhoneNumber(newContact.phoneNumber) == null) {
			contacts.add(newContact);
			System.out.println("Contact added successfully!");
		} else {
			System.out.println("Contact with this phone number already exists!");
		}
	}

	public void removeContact(String phoneNumber) {
		Contact toRemove = findByPhoneNumber(phoneNumber);
		if (toRemove != null) {
			contacts.remove(toRemove);
			System.out.println("Contact removed successfully!");
		} else {
			System.out.println("Contact not found!");
		}
	}

	public Contact findByPhoneNumber(String phoneNumber) {
		for (Contact contact : contacts) {
			if (contact.phoneNumber.equals(phoneNumber)) {
				return contact;
			}
		}
		return null;
	}

	public List<Contact> findByFirstName(String firstName) {
		List<Contact> results = new ArrayList<>();
		for (Contact contact : contacts) {
			if (contact.firstName.equalsIgnoreCase(firstName)) {
				results.add(contact);
			}
		}
		return results;
	}

	public List<Contact> findByLastName(String lastName) {
		List<Contact> results = new ArrayList<>();
		for (Contact contact : contacts) {
			if (contact.lastName.equalsIgnoreCase(lastName)) {
				results.add(contact);
			}
		}
		return results;
	}

	public void editContact(String phoneNumber, String newFirstName, String newLastName, String newPhoneNumber) {
		Contact toEdit = findByPhoneNumber(phoneNumber);
		if (toEdit != null) {
			toEdit.firstName = newFirstName;
			toEdit.lastName = newLastName;

			if (!phoneNumber.equals(newPhoneNumber)) {
				if (findByPhoneNumber(newPhoneNumber) == null) {
					toEdit.phoneNumber = newPhoneNumber;
				} else {
					System.out.println("New phone number already in use!");
				}
			}
			System.out.println("Contact updated successfully!");
			} else {
				System.out.println("Contact not found!");
		}
	}

	public void displayAllContacts() {
		if (contacts.isEmpty()) {
			System.out.println("No contacts available!");
			return;
		}

		System.out.println("All Contacts:");
		for (Contact contact : contacts) {
			System.out.println(contact.getFullName() + " - " + contact.phoneNumber);
		}
	}

	public static void main(String[] args) {
		ContactManager manager = new ContactManager();
		Scanner scanner = new Scanner(System.in);
 
		while (true) {
			System.out.println("\nContact Management System");
			System.out.println("""
			1. Add Contact
			2. Remove Contact
			3. Find by Phone Number
			4. Find by First Name
			5. Find by Last Name
			6. Edit Contact
			7. View All Contacts
			8. Exit
			""");
			System.out.print("Enter your choice: ");
			int choice = scanner.nextInt();
			scanner.nextLine();
			switch (choice) {
			case 1:
				System.out.print("Enter First Name: ");
				String firstName = scanner.nextLine();
				while(true) {
					if(firstName.trim().isEmpty()) {
						System.out.println("Name cannot be empty!");
					} 
					else if(!firstName.matches("[a-zA-Z]+")) {
						System.out.println("Name must contain only letters!");
					}
					else {
				break;
				}
					System.out.print("Enter First Name: ");
					firstName = scanner.nextLine();
				}
				System.out.print("Enter Last Name: ");
				String lastName = scanner.nextLine();
				while(true) {
					if(lastName.trim().isEmpty()) {
						System.out.println("Name cannot be empty!");
					} 
					else if(!lastName.matches("[a-zA-Z]+")) {
						System.out.println("Name must contain only letters!");
					}
					else {
				break;
				}
					System.out.print("Enter Last Name: ");
					lastName = scanner.nextLine();
				}
				System.out.print("Enter Phone Number: ");
				String phoneNumber;
				while (true) {
				phoneNumber = scanner.nextLine().trim();
				if (phoneNumber.matches("\\+?[0-9\\-\\s]{8,15}")) {
       					break;
				}
					System.out.println("Invalid phone number! Please enter 8-15 digits (may include +, - or spaces)");
					System.out.print("Enter Phone Number: ");
				}
				Contact newContact = new Contact(firstName, lastName, phoneNumber);
				manager.addContact(newContact);
			break;
                    
			case 2:
				System.out.print("Enter Phone Number to remove: ");
				String phoneToRemove;
				while (true) {
				phoneToRemove = scanner.nextLine().trim();
				if (phoneToRemove.matches("\\+?[0-9\\-\\s]{8,15}")) {
       					break;
				}
					System.out.println("Invalid phone number! Please enter 8-15 digits (may include +, - or spaces)");
					System.out.print("Enter Phone Number to remove: ");
				}

				manager.removeContact(phoneToRemove);
			break;
                    
			case 3:
				System.out.print("Enter Phone Number to search: ");
				String searchPhone;
				while (true) {
				searchPhone = scanner.nextLine().trim();
				if (searchPhone.matches("\\+?[0-9\\-\\s]{8,15}")) {
       					break;
				}
					System.out.println("Invalid phone number! Please enter 8-15 digits (may include +, - or spaces)");
					System.out.print("Enter Phone Number to search: ");
				}

				Contact foundByPhone = manager.findByPhoneNumber(searchPhone);
				if (foundByPhone != null) {
					System.out.println("Found: " + foundByPhone.getFullName());
				} else {
					System.out.println("Contact not found!");
				}
			break;
                    
			case 4:
				System.out.print("Enter First Name to search: ");
				String searchFirstName = scanner.nextLine();
				List<Contact> byFirstName = manager.findByFirstName(searchFirstName);
				displaySearchResults(byFirstName);
			break;
                    
			case 5:
				System.out.print("Enter Last Name to search: ");
				String searchLastName = scanner.nextLine();
				List<Contact> byLastName = manager.findByLastName(searchLastName);
				displaySearchResults(byLastName);
			break;
                    
			case 6:
				System.out.print("Enter Phone Number of contact to edit: ");
				String editPhone;
				while (true) {
				editPhone = scanner.nextLine().trim();
				if (editPhone.matches("\\+?[0-9\\-\\s]{8,15}")) {
       					break;
				}
					System.out.println("Invalid phone number! Please enter 8-15 digits (may include +, - or spaces)");
					System.out.print("Enter Phone Number of contact to edit: ");
				}

				System.out.print("Enter new First Name: ");
				String newFirstName = scanner.nextLine();
				while(true) {
					if(newFirstName.trim().isEmpty()) {
						System.out.println("Name cannot be empty!");
					} 
					else if(!newFirstName.matches("[a-zA-Z]+")) {
						System.out.println("Name must contain only letters!");
					}
					else {
				break;
				}
					System.out.print("Enter new First Name: ");
					newFirstName = scanner.nextLine();
				}
				System.out.print("Enter new Last Name: ");
				String newLastName = scanner.nextLine();
				while(true) {
					if(newLastName.trim().isEmpty()) {
						System.out.println("Name cannot be empty!");
					} 
					else if(!newLastName.matches("[a-zA-Z]+")) {
						System.out.println("Name must contain only letters!");
					}
					else {
				break;
				}
					System.out.print("Enter Last Name: ");
					newLastName = scanner.nextLine();
				}
				System.out.print("Enter new Phone Number: ");
				String newPhone;
				while (true) {
				newPhone = scanner.nextLine().trim();
				if (newPhone.matches("\\+?[0-9\\-\\s]{8,15}")) {
       					break;
				}
					System.out.println("Invalid phone number! Please enter 8-15 digits (may include +, - or spaces)");
					System.out.print("Enter new Phone Number: ");
				}
				manager.editContact(editPhone, newFirstName, newLastName, newPhone);
			break;
                    
			case 7:
				manager.displayAllContacts();
			break;
                    
			case 8:
				System.out.println("Exiting...");
			break;
                    
			default:
				System.out.println("Invalid choice!");
			}
		}
	}

	public static void displaySearchResults(List<Contact> results) {
		if (results.isEmpty()) {
			System.out.println("No contacts found!");
			return;
		}
        
		System.out.println("Found " + results.size() + " contacts:");
		for (Contact contact : results) {
			System.out.println(contact.getFullName() + " - " + contact.phoneNumber);
		}
	}

}

File: phonebook/contact_manager.js
class Contact {
	constructor(firstName, lastName, phoneNumber) {
		this.firstName = firstName;
		this.lastName = lastName;
		this.phoneNumber = phoneNumber;
	}

	getFullName() {
		return `${this.firstName} ${this.lastName}`;
	}

	toString() {
		return `Name: ${this.getFullName()}
	Phone: ${this.phoneNumber}`;
	}
}

class ContactManager {
	constructor() {
		this.contacts = [];
	}

  addContact(contact) {
    if (!this._phoneExists(contact.phoneNumber)) {
      this.contacts.push(contact);
      console.log("Contact added successfully!");
      return true;
    }
    console.log("Error: Phone number already exists!");
    return false;
  }

  removeContact(phoneNumber) {
    const index = this.contacts.findIndex(c => c.phoneNumber === phoneNumber);
    if (index !== -1) {
      this.contacts.splice(index, 1);
      console.log("Contact removed successfully!");
      return true;
    }
    console.log("Error: Contact not found!");
    return false;
  }

  findByPhone(phoneNumber) {
    return this.contacts.find(c => c.phoneNumber === phoneNumber);
  }

  findByFirstName(firstName) {
    return this.contacts.filter(c => 
      c.firstName.toLowerCase() === firstName.toLowerCase());
  }

  findByLastName(lastName) {
    return this.contacts.filter(c => 
      c.lastName.toLowerCase() === lastName.toLowerCase());
  }

  editContact(phoneNumber, updatedContact) {
    const index = this.contacts.findIndex(c => c.phoneNumber === phoneNumber);
    if (index !== -1) {
      if (updatedContact.phoneNumber !== phoneNumber && 
          this._phoneExists(updatedContact.phoneNumber)) {
        console.log("Error: New phone number already exists!");
        return false;
      }
      this.contacts[index] = updatedContact;
      console.log("Contact updated successfully!");
      return true;
    }
    console.log("Error: Contact not found!");
    return false;
  }

  displayAll() {
    if (this.contacts.length === 0) {
      console.log("No contacts available!");
      return;
    }
    console.log("\nAll Contacts:");
    console.log("-----------------------");
    this.contacts.forEach((contact, i) => {
      console.log(`${i+1}. ${contact.getFullName()} - ${contact.phoneNumber}`);
    });
    console.log("-----------------------");
  }

  _phoneExists(phoneNumber) {
    return this.contacts.some(c => c.phoneNumber === phoneNumber);
  }

  async run() {
    const readline = require('readline').createInterface({
      input: process.stdin,
      output: process.stdout
    });

    const question = (query) => new Promise(resolve => 
      readline.question(query, resolve));

    console.log("\n=== Contact Management System ===");

    while (true) {
      console.log("\nMain Menu:");
      console.log("1. Add Contact");
      console.log("2. Remove Contact");
      console.log("3. Search Contacts");
      console.log("4. Edit Contact");
      console.log("5. View All Contacts");
      console.log("6. Exit");

      const choice = await question("Enter your choice (1-6): ");

      switch (choice) {
        case '1':
          console.log("\nAdd New Contact");
          const firstName = await question("First Name: ");
          const lastName = await question("Last Name: ");
          const phoneNumber = await question("Phone Number: ");
          
          const newContact = new Contact(
            firstName.trim(), 
            lastName.trim(), 
            phoneNumber.trim(), 
          );
          this.addContact(newContact);
          break;

        case '2':
          console.log("\nRemove Contact");
          const phoneToRemove = await question("Enter phone number to remove: ");
          this.removeContact(phoneToRemove.trim());
          break;

        case '3':
          console.log("\nSearch Contacts");
          console.log("1. By Phone Number");
          console.log("2. By First Name");
          console.log("3. By Last Name");
          
          const searchChoice = await question("Choose search method (1-3): ");
          const searchTerm = await question("Enter search term: ");
          
          let results = [];
          switch (searchChoice) {
            case '1':
              const contact = this.findByPhone(searchTerm.trim());
              console.log("\nSearch Results:");
              console.log(contact ? contact.toString() : "No contact found!");
              break;
            case '2':
              results = this.findByFirstName(searchTerm.trim());
              this._displaySearchResults(results);
              break;
            case '3':
              results = this.findByLastName(searchTerm.trim());
              this._displaySearchResults(results);
              break;
            default:
              console.log("Invalid choice!");
          }
          break;

        case '4':
          console.log("\nEdit Contact");
          const oldPhone = await question("Enter phone number of contact to edit: ");
          const existing = this.findByPhone(oldPhone.trim());
          
          if (!existing) {
            console.log("Contact not found!");
            break;
          }
          
          console.log("\nCurrent contact details:");
          console.log(existing.toString());
          
          console.log("\nEnter new details (leave blank to keep current):");
          const newFirst = await question(`First Name [${existing.firstName}]: `) || existing.firstName;
          const newLast = await question(`Last Name [${existing.lastName}]: `) || existing.lastName;
          const newPhone = await question(`Phone [${existing.phoneNumber}]: `) || existing.phoneNumber;
          
          const updated = new Contact(
            newFirst.trim(),
            newLast.trim(),
            newPhone.trim()
          );
          this.editContact(oldPhone.trim(), updated);
          break;

        case '5':
          this.displayAll();
          break;

        case '6':
          console.log("\nThank you for using the Contact Management System. Goodbye!");
          readline.close();
          return;

        default:
          console.log("Invalid choice! Please enter a number between 1 and 6.");
      }
    }
  }

  _displaySearchResults(results) {
    if (results.length === 0) {
      console.log("No contacts found!");
      return;
    }
    console.log(`\nFound ${results.length} contact(s):`);
    console.log("-----------------------");
    results.forEach(contact => {
      console.log(contact.toString());
      console.log("-----------------------");
    });
  }
}


const manager = new ContactManager();
manager.run();